test/functional/testlib.bash

```diff
@@ -10,7 +11,7 @@ export TEST_NAME_SHORT
 export THEME_DIRNAME
 export FUNC_DIR
 export SWUPD_DIR="$FUNC_DIR/../.."
-export SWUPD="$SWUPD_DIR/swupd"
+export SWUPD="$TEST_ROOT_DIR/swupd"
```

Comment:
The problem is that for our bat test we need to assume where the swupd binary is. Currently we assume that the binary is always relative to the testlib location. This is correct for the way we build swupd on travis and how I build swupd for me, but this assumption is incorrect. In master if you run:

    autoreconf --force --install --symlink --warnings=all
    mkdir build
    cd build
    ../configure
    make -j
    make check #this will fail
    bats ../tests/functional/bundleremove/ #this will also fail

So current testlib just work if the swupd binary is located in the path that is ../../ relative to the testlib directory. If you want to save the build artifacts in another directory tests are going to fail.

My patch is changing the assumption to assume the binary is in the current directory where the person is running the tests. This is not perfect, but in my opinion is more standard. Another option we can have is to add a new environment variable to set this. Besides that I don't see a better solution to the problem.

Do you think that the use case when a person is running the test in a different directory should be supported? Is setting an environment variable a good solution for your case?

Reply:
What about adding something like this to the top of testlib.bats to auto-detect the binary:

```
#!/usr/bin/bash

FUNC_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
TEST_FILENAME=$(basename "$BATS_TEST_FILENAME")
TEST_NAME=${TEST_FILENAME%.bats}
THEME_DIRNAME="$BATS_TEST_DIRNAME"

export TEST_NAME
export TEST_NAME_SHORT
export THEME_DIRNAME
export FUNC_DIR
export SWUPD_DIR="$FUNC_DIR/../.."
export SWUPD_CONFIG_DIR="$SWUPD_DIR"/testconfig
export SWUPD_CONFIG_FILE="$SWUPD_CONFIG_DIR"/config

# detect where the swupd binary is
if [ -e "$SWUPD" ]; then
	# nothing to be done, variable already set up
        :
elif [ -e "$SWUPD_DIR"/swupd ]; then
	# using the path relative to the test dir
	export SWUPD="$SWUPD_DIR"/swupd
elif [ -e "$(pwd)"/swupd ]; then
	# using the current path
	export SWUPD="$(pwd)"/swupd
fi
```

Reply:
good idea. I'll use that